### DISCUSSION TITLE
Request Enhancement:  New define for colorspace when doing -compose colorize

### DISCUSSION THREAD

**Discussion by fmw42** [ImageMagick/ImageMagick#7340] [Development]
### ImageMagick version

7.1.1.32

### Operating system, version and so on

Mac OSX Ventura

### Description

Please add a new define something like

`-define compose:colorspace=XXX  or -define compose:colorize-colorspace=XXX`

to be used when doing

`magick image1 image2 -compose colorize -composite result`

So that different colorspace might be used when doing the colorize composite.

**Comment by dlemstra** [comment]
Could you not change the colorspace of the image instead?

**Reply by fmw42** [reply]
I do not think that is the same thing.  

These are not the same.  -compose colorize is supposed to change colorspace to one of the Hue colorspaces such as HSV, then swap the H channel between the two images.

`magick img.png \( +clone -fill red -colorize 100 \) -compose colorize -composite x.png`

So this emulates that process for an image with transparency, but they are not the same.

```
magick img.png \
\( -clone 0 -alpha extract -write mpr:alpha +delete \) \
\( -clone 0 -alpha off -fill red -colorize 100 -colorspace HSV -separate -delete -1,-2 \) \
\( -clone 0 -alpha off -colorspace HSV -separate -delete -3 \) \
-delete 0 -set colorspace HSV -combine -colorspace sRGB \
mpr:alpha -alpha off -compose copy_opacity -composite \
result.png 
```

I have tried using HSV, HSL, HCL and not reproduce the -compose colorize.


I note that we have -define modulate:colorspace=XXX to affect what colorspace is used for -modulate.  I am asking for something similar for -compose colorize.

**Comment by urban-warrior** [comment]
Image composition closely follows the SVG compositing standard.  As such it requires all operations to take place in the sRGB color space.   Instead, convert the colorspace but declare it sRGB so compositing does not transform a non-sRGB colorspace:

```
magick -colorspace HCL -set colorspace sRGB image1 image2 -compose colorize -composite -set colorspace HCL -colorspace sRGB  result
```

**Reply by fmw42** [reply]
Was HCL the colorspace used in -compose colorize? Please confirm.

 I will give that a try tomorrow and see if my longer attempt matches the simpler one above.

**Reply by urban-warrior** [reply]
Yes, its HCL.  After analyzing the algorithm, it turns out that we can support alternative colorspaces.  We are pushing a release today.  After the release, we'll work on a solution to this enhancement request.

**Reply by fmw42** [reply]
Thanks. It is not urgent. But is appreciated

**Reply by urban-warrior** [reply]
Try `-define compose:colorspace` in about an hour.

**Reply by fmw42** [reply]
Thanks it seems to be working and reproduces the default HCL colorspace and gives different results for other colorspaces as expected.

**Comment by fmw42** [comment]
This is the result of using the new (beta) to process the following image:

![img](https://github.com/ImageMagick/ImageMagick/assets/13458218/a42ba76e-5d6d-4338-83ed-ad2e3b2db9bd)

`magick img.png \( +clone -fill red -colorize 100 \) -compose colorize -composite result6b.png`

![result6b](https://github.com/ImageMagick/ImageMagick/assets/13458218/981840d9-c309-4644-ba32-303c17796588)

and is identical to 

`magick img.png \( +clone -fill red -colorize 100 \) -define compose:colorspace=HCL -compose colorize -composite result6bb.png`

![result6bb](https://github.com/ImageMagick/ImageMagick/assets/13458218/49cd0c5d-17d3-4bb0-9baf-70039c3526d6)


But I cannot reproduce with a swapping of hue channels

```
magick img.png \
\( -clone 0 -alpha extract -write mpr:alpha +delete \) \
\( -clone 0 -alpha off -colorspace HCL -set colorspace sRGB \) \
\( -clone 0 -alpha off -fill red -colorize 100 -colorspace HCL -set colorspace sRGB \) \
-delete 0 \
-compose colorize -composite -set colorspace HCL -colorspace sRGB  \
mpr:alpha -alpha off -compose copy_opacity -composite \
result2.png
```

![result2](https://github.com/ImageMagick/ImageMagick/assets/13458218/f28f686f-4d94-44df-8d20-d43c4fcc2517)

which produces the same result using the more simple

```
magick \
\( img.png -colorspace HCL \) \
\( -clone 0 -colorspace sRGB -fill red -colorize 100 -colorspace HCL \)  \
-compose colorize -composite -set colorspace HCL -colorspace sRGB  result3.png
```

![result3](https://github.com/ImageMagick/ImageMagick/assets/13458218/3d41dc40-2689-4821-bea6-42bafdb2fbd2)


However, removing the -set colorspace sRGB in the following results in a closer but not identical result

```
magick img.png \
\( -clone 0 -alpha extract -write mpr:alpha +delete \) \
\( -clone 0 -alpha off -fill red -colorize 100 -colorspace HCL -separate \) \
\( -clone 0 -alpha off -colorspace HCL -separate \) \
-delete 0,2,3,4 -set colorspace HCL -combine -colorspace sRGB \
mpr:alpha -alpha off -compose copy_opacity -composite \
result4.png 
```

![result4](https://github.com/ImageMagick/ImageMagick/assets/13458218/a8fda596-ba51-490a-9221-3930f217922c)


**@urban-warrior. What is being done in -compose colorize that I am missin**g

**Comment by urban-warrior** [comment]
The formula is to extract H & C from source and L from the composite image.  We then convert the HCL to sRGB.

**Reply by fmw42** [reply]
Thanks. I thought it was just the hue being swapped.  My mistake. So it is the same as Luminize, but with the images swapped. (Going back to Anthony's pages, I see he notes that. I should have re-read that page. Sorry)

This works to match

```
magick img.png \
\( -clone 0 -alpha extract -write mpr:alpha +delete \) \
\( -clone 0 -alpha off -fill red -colorize 100 -colorspace HCL -separate \) \
\( -clone 0 -alpha off -colorspace HCL -separate \) \
-delete 0,3,4,5 -set colorspace HCL -combine -colorspace sRGB \
mpr:alpha -alpha off -compose copy_opacity -composite \
result9.png 
```

![result9](https://github.com/ImageMagick/ImageMagick/assets/13458218/d6d72135-5ab7-483a-9ccb-f465795764bf)
